// syncano/lib/models/class.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ClassMeta = (0, _base.Meta)({
  name: 'class',
  pluralName: 'classes',
  endpoints: {
    'detail': {
      'methods': ['delete', 'patch', 'put', 'get'],
      'path': '/v1.1/instances/{instanceName}/classes/{name}/'
    },
    'list': {
      'methods': ['post', 'get'],
      'path': '/v1.1/instances/{instanceName}/classes/'
    }
  },
  relatedModels: ['DataObject']
});

var ClassConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    }
  },
  name: {
    presence: true,
    string: true
  },
  description: {
    string: true
  },
  schema: {
    array: true
  },
  group: {
    numericality: {
      noStrings: true
    }
  },
  group_permissions: {
    inclusion: ['none', 'read', 'create_objects']
  },
  other_permissions: {
    inclusion: ['none', 'read', 'create_objects']
  },
  metadata: {
    object: true
  }
};

/**
 * OO wrapper around instance groups {@link http://docs.syncano.com/v4.0/docs/instancesinstanceclasses endpoint}.
 * @constructor
 * @type {Class}

 * @property {String} name
 * @property {String} instanceName
 * @property {Number} objects_count
 * @property {Array} schema
 * @property {String} status
 * @property {Object} metadata
 * @property {String} revision
 * @property {String} expected_revision
 * @property {String} group
 * @property {String} group_permissions
 * @property {String} other_permissions
 * @property {String} [description = null]
 * @property {String} [links = {}]
 * @property {Date} [created_at = null]
 * @property {Date} [updated_at = null]

 */
var Class = (0, _stampit2.default)().compose(_base.Model).setMeta(ClassMeta).setConstraints(ClassConstraints);

exports.default = Class;
module.exports = exports['default'];
