// syncano/lib/models/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _instance = require('./instance');

var _instance2 = _interopRequireDefault(_instance);

var _class = require('./class');

var _class2 = _interopRequireDefault(_class);

var _channel = require('./channel');

var _channel2 = _interopRequireDefault(_channel);

var _dataobject = require('./dataobject');

var _dataobject2 = _interopRequireDefault(_dataobject);

var _user = require('./user');

var _user2 = _interopRequireDefault(_user);

var _group = require('./group');

var _group2 = _interopRequireDefault(_group);

var _admin = require('./admin');

var _admin2 = _interopRequireDefault(_admin);

var _apikey = require('./apikey');

var _apikey2 = _interopRequireDefault(_apikey);

var _instanceinvitation = require('./instanceinvitation');

var _instanceinvitation2 = _interopRequireDefault(_instanceinvitation);

var _invitation = require('./invitation');

var _invitation2 = _interopRequireDefault(_invitation);

var _invoice = require('./invoice');

var _invoice2 = _interopRequireDefault(_invoice);

var _script = require('./script');

var _script2 = _interopRequireDefault(_script);

var _schedule = require('./schedule');

var _schedule2 = _interopRequireDefault(_schedule);

var _trigger = require('./trigger');

var _trigger2 = _interopRequireDefault(_trigger);

var _scriptendpoint = require('./scriptendpoint');

var _scriptendpoint2 = _interopRequireDefault(_scriptendpoint);

var _dataendpoint = require('./dataendpoint');

var _dataendpoint2 = _interopRequireDefault(_dataendpoint);

var _scripttrace = require('./scripttrace');

var _scripttrace2 = _interopRequireDefault(_scripttrace);

var _scheduletrace = require('./scheduletrace');

var _scheduletrace2 = _interopRequireDefault(_scheduletrace);

var _triggertrace = require('./triggertrace');

var _triggertrace2 = _interopRequireDefault(_triggertrace);

var _scriptendpointtrace = require('./scriptendpointtrace');

var _scriptendpointtrace2 = _interopRequireDefault(_scriptendpointtrace);

var _gcmdevice = require('./gcmdevice');

var _gcmdevice2 = _interopRequireDefault(_gcmdevice);

var _gcmconfig = require('./gcmconfig');

var _gcmconfig2 = _interopRequireDefault(_gcmconfig);

var _apnsdevice = require('./apnsdevice');

var _apnsdevice2 = _interopRequireDefault(_apnsdevice);

var _apnsconfig = require('./apnsconfig');

var _apnsconfig2 = _interopRequireDefault(_apnsconfig);

var _gcmmessage = require('./gcmmessage');

var _gcmmessage2 = _interopRequireDefault(_gcmmessage);

var _apnsmessage = require('./apnsmessage');

var _apnsmessage2 = _interopRequireDefault(_apnsmessage);

var _template = require('./template');

var _template2 = _interopRequireDefault(_template);

var _fullBackup = require('./fullBackup');

var _fullBackup2 = _interopRequireDefault(_fullBackup);

var _partialBackup = require('./partialBackup');

var _partialBackup2 = _interopRequireDefault(_partialBackup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  Instance: _instance2.default,
  Class: _class2.default,
  Channel: _channel2.default,
  DataObject: _dataobject2.default,
  User: _user2.default,
  Group: _group2.default,
  Admin: _admin2.default,
  ApiKey: _apikey2.default,
  InstanceInvitation: _instanceinvitation2.default,
  Invitation: _invitation2.default,
  Invoice: _invoice2.default,
  Script: _script2.default,
  Schedule: _schedule2.default,
  Trigger: _trigger2.default,
  ScriptEndpoint: _scriptendpoint2.default,
  DataEndpoint: _dataendpoint2.default,
  ScriptTrace: _scripttrace2.default,
  ScheduleTrace: _scheduletrace2.default,
  TriggerTrace: _triggertrace2.default,
  ScriptEndpointTrace: _scriptendpointtrace2.default,
  GCMDevice: _gcmdevice2.default,
  APNSDevice: _apnsdevice2.default,
  APNSConfig: _apnsconfig2.default,
  GCMMessage: _gcmmessage2.default,
  GCMConfig: _gcmconfig2.default,
  APNSMessage: _apnsmessage2.default,
  Template: _template2.default,
  FullBackup: _fullBackup2.default,
  PartialBackup: _partialBackup2.default
};
module.exports = exports['default'];
