// syncano/lib/models/invitation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _querySet = require('../querySet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var InvitationQuerySet = (0, _stampit2.default)().compose(_querySet.BaseQuerySet, _querySet.Create, _querySet.BulkCreate, _querySet.Get, _querySet.GetOrCreate, _querySet.Delete, _querySet.List, _querySet.Delete).methods({
  accept: function accept(invitation_key) {
    this.method = 'POST';
    this.endpoint = 'accept';
    this.payload = { invitation_key: invitation_key };
    this._serialize = false;

    return this;
  }
});

var InvitationMeta = (0, _base.Meta)({
  name: 'invitation',
  pluralName: 'invitations',
  endpoints: {
    'detail': {
      'methods': ['delete', 'get'],
      'path': '/v1.1/account/invitations/{id}/'
    },
    'list': {
      'methods': ['post', 'get'],
      'path': '/v1.1/account/invitations/'
    },
    'accept': {
      'methods': ['post'],
      'path': '/v1.1/account/invitations/accept/'
    }
  }
});

/**
 * OO wrapper around invitations {@link # endpoint}.
 * @constructor
 * @type {Invitation}

 * @property {String} email
 * @property {String} role
 * @property {String} [key = null]
 * @property {String} [inviter = null]
 * @property {String} [status = null]
 * @property {Number} [id = null]
 * @property {String} [description = null]
 * @property {String} [links = {}]
 * @property {Date} [created_at = null]
 * @property {Date} [updated_at = null]
 */
var Invitation = (0, _stampit2.default)().compose(_base.Model).setMeta(InvitationMeta).setQuerySet(InvitationQuerySet).methods({
  accept: function accept(invitation_key) {
    var meta = this.getMeta();
    var path = meta.resolveEndpointPath('accept', this);

    return this.makeRequest('POST', path, { invitation_key: invitation_key });
  }
});

exports.default = Invitation;
module.exports = exports['default'];
