// syncano/lib/models/partialBackup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _querySet = require('../querySet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PartialBackupQuerySet = (0, _stampit2.default)().compose(_querySet.BaseQuerySet, _querySet.List, _querySet.Get, _querySet.Delete, _querySet.Create);

var PartialBackupMeta = (0, _base.Meta)({
  name: 'partialBackup',
  pluralName: 'partialBackups',
  endpoints: {
    'detail': {
      'methods': ['delete', 'get'],
      'path': '/v1.1/instances/{instanceName}/backups/partial/{id}/'
    },
    'list': {
      'methods': ['get', 'post'],
      'path': '/v1.1/instances/{instanceName}/backups/partial/'
    }
  }
});

var PartialBackupConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    }
  },
  description: {
    string: true
  },
  label: {
    string: true
  },
  query_args: {
    presence: true,
    object: true
  }
};

/**
 * OO wrapper around instance backups {@link # endpoint}.
 * @constructor
 * @type {Backup}

 * @property {Number} id
 * @property {String} instance
 * @property {String} created_at
 * @property {String} updated_at
 * @property {String} archive - link with zipped backup file
 * @property {Number} size
 * @property {String} status
 * @property {String} status_info
 * @property {String} description
 * @property {String} label
 * @property {Object} [links = {}]
 */
var PartialBackup = (0, _stampit2.default)().compose(_base.Model).setQuerySet(PartialBackupQuerySet).setMeta(PartialBackupMeta).setConstraints(PartialBackupConstraints);

exports.default = PartialBackup;
module.exports = exports['default'];
