// syncano/lib/models/scripttrace.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _querySet = require('../querySet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ScriptTraceQuerySet = (0, _stampit2.default)().compose(_querySet.BaseQuerySet, _querySet.Get, _querySet.List);

var ScriptTraceMeta = (0, _base.Meta)({
  name: 'scripttrace',
  pluralName: 'scripttrace',
  endpoints: {
    'detail': {
      'methods': ['delete', 'patch', 'put', 'get'],
      'path': '/v1.1/instances/{instanceName}/snippets/scripts/{scriptId}/traces/{id}/'
    },
    'list': {
      'methods': ['post', 'get'],
      'path': '/v1.1/instances/{instanceName}/snippets/scripts/{scriptId}/traces/'
    }
  }
});

var ScriptConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    }
  },
  scriptId: {
    presence: true
  }
};

/**
 * OO wrapper around script trace {@link # endpoint}.
 * This model is *read only*.
 * @constructor
 * @type {ScriptTrace}

 * @property {Number} id
 * @property {String} instanceName
 * @property {Number} scriptId
 * @property {String} status
 * @property {Date} executed_at
 * @property {Number} duration
 * @property {Object} [result = {}]
 * @property {String} result.stderr
 * @property {String} result.stdout
 * @property {String} [links = {}]
 */
var ScriptTrace = (0, _stampit2.default)().compose(_base.Model).setQuerySet(ScriptTraceQuerySet).setConstraints(ScriptConstraints).setMeta(ScriptTraceMeta);

exports.default = ScriptTrace;
module.exports = exports['default'];
