// syncano/lib/models/scriptendpointtrace.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _querySet = require('../querySet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ScriptEndpointTraceQuerySet = (0, _stampit2.default)().compose(_querySet.BaseQuerySet, _querySet.Get, _querySet.List);

var ScriptEndpointTraceMeta = (0, _base.Meta)({
  name: 'triggertrace',
  pluralName: 'triggertraces',
  endpoints: {
    'detail': {
      'methods': ['get'],
      'path': '/v1.1/instances/{instanceName}/endpoints/scripts/{scriptEndpointName}/traces/{id}/'
    },
    'list': {
      'methods': ['get'],
      'path': '/v1.1/instances/{instanceName}/endpoints/scripts/{scriptEndpointName}/traces/'
    }
  }
});

var ScriptEndpointTraceConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    }
  },
  scriptEndpointName: {
    presence: true,
    string: true
  }
};

/**
 * OO wrapper around webhook traces {@link # endpoint}.
 * This model is *read only*.
 * @constructor
 * @type {ScriptEndpointTrace}

 * @property {Number} id
 * @property {String} instanceName
 * @property {String} scriptEndpointName
 * @property {String} status
 * @property {Date} executed_at
 * @property {Number} duration
 * @property {Object} [result = {}]
 * @property {String} result.stderr
 * @property {String} result.stdout
 * @property {String} [links = {}]
 */
var ScriptEndpointTrace = (0, _stampit2.default)().compose(_base.Model).setMeta(ScriptEndpointTraceMeta).setQuerySet(ScriptEndpointTraceQuerySet).setConstraints(ScriptEndpointTraceConstraints);

exports.default = ScriptEndpointTrace;
module.exports = exports['default'];
