// syncano/lib/models/trigger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TriggerMeta = (0, _base.Meta)({
  name: 'trigger',
  pluralName: 'triggers',
  endpoints: {
    'detail': {
      'methods': ['delete', 'patch', 'put', 'get'],
      'path': '/v1.1/instances/{instanceName}/triggers/{id}/'
    },
    'list': {
      'methods': ['post', 'get'],
      'path': '/v1.1/instances/{instanceName}/triggers/'
    }
  },
  relatedModels: ['TriggerTrace']
});

var TriggerConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    }
  },
  label: {
    presence: true,
    string: true
  },
  description: {
    string: true
  },
  signal: {
    presence: true,
    inclusion: ['post_update', 'post_create', 'post_delete']
  },
  script: {
    presence: true,
    numericality: {
      noStrings: true
    }
  },
  class: {
    presence: true,
    string: true
  }
};

/**
 * OO wrapper around instance triggers {@link # endpoint}.
 * @constructor
 * @type {Trigger}

 * @property {Number} id
 * @property {String} instanceName
 * @property {String} label
 * @property {String} signal
 * @property {Number} codebox
 * @property {String} class
 * @property {String} [description = null]
 * @property {String} [links = {}]
 * @property {Date} [created_at = null]
 * @property {Date} [updated_at = null]
 */
var Trigger = (0, _stampit2.default)().compose(_base.Model).setConstraints(TriggerConstraints).setMeta(TriggerMeta);

exports.default = Trigger;
module.exports = exports['default'];
